### PR TITLE
Make Matrix event cache recovery fail closed

### PR DESCRIPTION
**Pull request by basnijholt**
## What changed

- invalidate every affected room before writing a limited `/sync` timeline, reset active and persisted sync positions in every trust state, and allow only a later complete response to establish a new certified checkpoint
- treat opaque `m.room.encrypted` timeline and history evidence as unknown thread impact, preserving stale snapshots until a decryption-capable refresh reconstructs complete history
- prevent opaque ciphertext from overwriting a decrypted point-event payload while allowing clear normalized improvements, with atomic derived-index reconciliation in SQLite and PostgreSQL
- classify relation cycles and resolver hop-cap exhaustion as indeterminate instead of room-level
- persist known outbound room-level messages, edits, and redactions through the ordered cache coordinator, including same-turn read-your-write and read-your-redaction behavior

## Root cause

The cache treated partial timelines, undecryptable events, and unresolved relation chains as sufficiently classified to update or certify reconstructed history. Point-event upserts also replaced payloads unconditionally, and outbound room-level mutations skipped the cache boundary. Together these paths could make an incomplete or stale local view appear authoritative.

## Impact

Matrix cache snapshots now remain advisory and stale whenever completeness cannot be proved. A complete non-limited recovery can certify a new checkpoint, decrypted payload quality is monotonic across bots/backends, and successful room-level mutations are immediately observable without breaking sync-echo idempotence.

## Validation

- `uv sync --all-extras`
- focused Matrix cache suites, including disposable PostgreSQL parity for both payload arrival orders, duplicate IDs in one batch, relation moves, outbound send/edit/redaction ordering, and same-turn reads
- `uv run pytest` (`10730 passed, 151 skipped` on the latest-main merge)
- `uv run pre-commit run --all-files`
- `uv run tach check --dependencies --interfaces`
- independent native zero-tolerance pre-PR review and follow-up re-review: approved with no remaining blocker

## Summary by CodeRabbit

* **New Features**
  * Added a “wait until room cache is idle” capability to prevent stale reads during ongoing cache writes.
* **Bug Fixes**
  * Improved caching for opaque encrypted events: decrypted content is preferred, and encrypted payloads no longer overwrite already-resolved clear data.
  * Hardened thread history/snapshots and sync certification with fail-closed behavior for opaque-encrypted and limited-sync scenarios, including safer token-reset semantics.
  * Fixed outbound message/edit/redaction cache bookkeeping so point reads and conversation history stay consistent.
  * Updated thread relation resolution to fail safe (indeterminate) on cycles and excessive hop counts.
  * Enhanced bulk cache persistence, derived index maintenance, and cleanup for events and thread indexes.
* **Documentation**
  * Updated the Matrix event-cache interaction contract to reflect the new encrypted and limited-sync behaviors.
* **Tests**
  * Expanded coverage for opaque-encrypted, limited-sync, and certification/invalidation edge cases.

### CODE REVIEW

File: src/mindroom/matrix/cache/thread_writes.py [hunk 1]
```diff
@@ -1137,14 +1178,22 @@ def cache_sync_timeline(
...
         """Queue sync timeline persistence through the room-ordered cache barrier."""
         if not self._cache_ops.cache_runtime_available():
             return []
+        if limited_room_ids is None:
+            limited_room_ids, validation_errors = self._limited_sync_timeline_room_ids(response)
+            if validation_errors:
+                if raise_on_cache_write_failure:
+                    raise validation_errors[0]
+                return []
         room_plain_events, room_threaded_events, room_redactions = self._group_sync_timeline_updates(response)
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/matrix/cache/thread_writes.py]: Verified against the runtime path: production sync uses cache_sync_timeline_for_certification(), which returns validation failures in SyncCacheWriteResult.errors; AgentBot then emits matrix_sync_certification_uncertain with those error details. The fail-open direct helper has no production caller, so an extra warning here would duplicate no runtime diagnostic. No code change.

File: src/mindroom/matrix/conversation_cache.py [hunk 2]
```diff
@@ -494,6 +494,9 @@ async def get_event(
     ) -> EventLookupResult:
         """Resolve one event through per-turn memoization and the advisory cache."""
         normalized_event_id = event_id.strip()
+        coordinator = self.runtime.event_cache_write_coordinator
+        if coordinator is not None and not coordinator.current_task_holds_room_barrier(room_id):
+            await coordinator.wait_for_room_idle(room_id)
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/matrix/conversation_cache.py]: Verified the coordinator lifecycle. This wait is the strict write-read barrier required for read-your-write and read-your-redaction; queued tasks wake waiters on completion, failure, or cancellation. Timing out and continuing would permit stale reads. The 5 s close timeout is shutdown policy, not a correctness-preserving read policy. A bounded point-read mode would need to fail closed and is a separate product decision, so no code change here.

File: src/mindroom/matrix/thread_membership.py [hunk 3]
```diff
@@ -174,6 +175,8 @@ def _next_related_event_target(
     current_event_id: str,
 ) -> str | None:
     """Return the next related event to inspect."""
+    if event_info.next_related_event_id("") == current_event_id:
+        return current_event_id
     return event_info.next_related_event_id(current_event_id)
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/matrix/thread_membership.py]: Verified: the empty string is the established no-exclusion argument throughout this resolver, and the direct self-cycle regression test covers this branch. Adding a one-off explanatory comment would repeat the helper contract without improving correctness, so no code change.

File: src/mindroom/matrix/cache/sqlite_event_cache_events.py [hunk 4]
```diff
@@ -294,28 +294,72 @@ async def write_lookup_index_rows(
...
             (
                 event.event_id,
                 room_id,
                 event.origin_server_ts,
                 event.event_json,
                 cached_at,
+            ),
+        )
+        accepted_row = await cursor.fetchone()
+        await cursor.close()
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/matrix/cache/sqlite_event_cache_events.py]: Fixed in 08f5d6462. The per-event cursor path is gone: SQLite now performs one set-based JSON-backed conditional upsert and consumes RETURNING inside an async context manager, so cursor cleanup is exception-safe.

File: src/mindroom/matrix/cache/sqlite_event_cache_events.py [hunk 5]
```diff
@@ -294,28 +294,72 @@ async def write_lookup_index_rows(
...
+            cursor = await db.execute(
+                """
+                SELECT room_id, thread_id
+                FROM event_threads
+                WHERE event_id = ? AND event_id <> thread_id
+                """,
+                (event.event_id,),
+            )
+            prior_thread_row = await cursor.fetchone()
+            await cursor.close()
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/matrix/cache/sqlite_event_cache_events.py]: Fixed in 08f5d6462. The per-event SELECT cursor is gone; prior thread roots are fetched set-wise with an async context manager, guaranteeing cursor cleanup.

File: src/mindroom/matrix/cache/sqlite_event_cache_events.py [hunk 6]
```diff
@@ -294,28 +294,72 @@ async def write_lookup_index_rows(
...
+            if prior_thread_row is not None:
+                prior_child_roots.add((str(prior_thread_row[0]), str(prior_thread_row[1])))
+    accepted_events = list(accepted_events_by_id.values())
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/matrix/cache/sqlite_event_cache_events.py]: Fixed in 08f5d6462 at the two owning backend seams. Duplicate IDs are reduced with the same clear-over-opaque/null semantics as SQL, payload acceptance remains atomic via conditional bulk UPSERT ... RETURNING, and all index cleanup/upserts are set-based. Regression tests cover duplicate orders, rejected opaque batches, missing/null types, payload/index parity on real SQLite and PostgreSQL, plus constant statement counts for 50-event batches. The cache-layer suite passes 150/150.

**Comment by basnijholt** [comment]
Addressed the current validation-skip/YAGNI review in c2614c907. `cache_sync_timeline` now validates every invocation and raises malformed responses instead of carrying an optional test-only prevalidated branch. The same cleanup removes unused exports/return values and coordinator states that current scheduler invariants cannot produce. Bulk opaque history now carries one explicit room-scan uncertainty flag instead of copying every ciphertext event into every requested thread. Net production change: 55 lines removed. Full pytest (10,751 passed), pre-commit, and Tach pass.

File: src/mindroom/matrix/cache/event_cache_events.py [hunk 7]
```diff
@@ -83,6 +83,21 @@ def serialize_cacheable_events(cacheable_events: list[_CachedEventValue]) -> lis
...
+    preferred_events: dict[str, SerializedCachedEvent] = {}
+    for event in events:
+        previous = preferred_events.get(event.event_id)
+        event_type = event.event.get("type")
+        if (
+            previous is None
+            or previous.event.get("type") == "m.room.encrypted"
+            or (event_type is not None and event_type != "m.room.encrypted")
+        ):
+            preferred_events[event.event_id] = event
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/matrix/cache/event_cache_events.py]: Fixed in fca75e338. In-memory duplicate reduction plus SQLite and PostgreSQL point/snapshot/append upserts now require a non-null incoming event type before an encrypted payload can be replaced. Focused SQLite/PostgreSQL tests and the full 10,820-test suite pass.

File: tests/test_event_cache.py [hunk 8]
```diff
@@ -1455,6 +1457,418 @@ async def test_individual_event_cache_store_and_retrieve(event_cache: Conversati
...
+@pytest.mark.asyncio
+@pytest.mark.parametrize(
+    ("payload_order", "expected_payload"),
+    [
+        (("opaque-a", "clear-a", "opaque-b"), "clear-a"),
+        (("clear-a", "clear-b"), "clear-b"),
+        (("opaque-a", "opaque-b"), "opaque-b"),
+        (("clear-a", "missing-type"), "clear-a"),
+        (("clear-a", "null-type"), "clear-a"),
+    ],
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply tests/test_event_cache.py:1575]: Fixed in fca75e338. The batch payload-order matrix includes opaque followed by missing type and opaque followed by null type, both preserving the opaque payload and its indexes. 1d8dfb6d2 removes only redundant lower-seam duplication; both backends remain covered. Full suite passes.

**Comment by basnijholt** [comment]
Addressed the latest CodeRabbit review in fca75e338, 1d8dfb6d2, and 43429287f. Missing/null-type payloads can no longer replace ciphertext in the in-memory reducer or SQLite/PostgreSQL point, snapshot, and append upserts; exact payload-order regressions cover both backends. The tautological relation-target helper is removed while preserving self-cycle behavior through the established no-exclusion call. Full pytest: 10,820 passed, 151 skipped. Pre-commit and Tach pass.

File: src/mindroom/matrix/cache/event_cache_events.py (hunk 7 above)
[the remark this answers is not in the document]
Comment (basnijholt) [reply src/mindroom/matrix/cache/event_cache_events.py]: Not changed after a production-reachability audit. Every production cache writer receives a Matrix PDU or a locally normalized outbound event with a required type; typeless payloads are injected only by direct low-level storage tests. Adding equivalent guards at six SQL sites would create an unreachable defensive branch. The production invariant remains clear/decrypted versus m.room.encrypted, covered in both arrival orders on SQLite and PostgreSQL.

File: tests/test_event_cache.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (basnijholt) [reply tests/test_event_cache.py:1575]: Not added after a production-reachability audit. Missing or null event types do not reach this storage path from Matrix sync, Matrix point reads, thread refill, or normalized outbound writes. These cases would test an invalid low-level injection and require permanent production guards for a state the application cannot produce.

**Comment by basnijholt** [comment]
Independent Claude Code Fable 5 review completed against `43429287f` in an isolated agent-cli worktree.

It reported two blockers, both verified and fixed in `fb0a732fb`:

1. `event_thread_rows()` could emit two rows for the same `(room_id, event_id)` when an event was both a thread child and another event's referenced root.
   Real PostgreSQL raised `CardinalityViolation`; SQLite silently chose the synthesized self-row.
   The shared row builder now deduplicates by the backend conflict key and gives an event's own mapping precedence.
   A real SQLite/PostgreSQL regression covers `$nested -> $child`, `$child -> $parent`.
2. `ThreadMutationCacheOps.store_events_batch()` contained a behaviorally redundant call-shape branch.
   It now always passes the typed `thread_id` keyword, and strict fakes implement the declared protocol.

The remote branch advanced during that review with the opaque-relation scoping work through `2eaaea0d9`.
I merged it normally, verified its relationless-ciphertext premise against the Matrix relationship contract, retained its valid scoped invalidation behavior, and caught one merge regression: the remote resolution had removed the typeless-payload downgrade guards and two point-cache regression cases.
`6053432a4` restores the shared reducer plus SQLite/PostgreSQL point, replace, and append guards and the point/thread regression matrix.

Final local validation at pushed head `6053432a4`:

- focused combined cache/sync/history/contract suites: passed on real SQLite and PostgreSQL;
- `uv run tach check --dependencies --interfaces`: passed;
- `uv run pre-commit run --all-files`: all hooks passed;
- `uv run pytest`: **10,824 passed, 151 skipped**.

No merge performed.

**Comment by basnijholt** [comment]
Follow-up feedback triage correction:

The typeless-payload guards mentioned in my prior status comment are **not** part of the final tree.
The original CodeRabbit finding was later withdrawn after a production-reachability audit.
I independently repeated that audit across every cache writer: production inputs are typed Matrix PDUs or locally constructed typed events; missing/null types occur only through direct malformed low-level test injection.
Keeping one reducer guard, six SQL branches, and invalid-input regression cases would therefore preserve unreachable defensive behavior, contrary to this repository's simplicity policy.

The final tree keeps:

- the real Fable PostgreSQL/SQLite duplicate thread-index key fix and dual-backend regression;
- the typed `store_events_batch(..., thread_id=...)` simplification;
- the merged scoped opaque-relation policy;
- clear/decrypted versus `m.room.encrypted` monotonicity in both arrival orders.

The final full gate was run on this exact tree before the history-only merge commit:

- Tach: passed;
- all pre-commit hooks: passed;
- pytest: **10,816 passed, 151 skipped**.

Head history was reconciled without force-push; the last merge commit is tree-identical to the fully tested parent.

File: docs/dev/matrix-event-cache-interaction-contract.md [hunk 9]
```diff
@@ -91,8 +117,12 @@ A durable thread snapshot is usable only when its state row exists, `validated_a
...
 
+A snapshot containing any still-opaque encrypted event is rejected and marked stale.
+
 A rejected or absent snapshot causes an authoritative homeserver room-history scan and guarded cache refill.
 
+A reconstructed thread containing opaque encrypted evidence cannot replace the prior snapshot and remains stale until a decryption-capable scan reconstructs complete history.
+
 A second unchanged read is served from cache and performs no homeserver scan.
 
 The advisory read path may use a labelled stale-cache fallback when a required refill fails.
```
[the remark this answers is not in the document]
Comment (basnijholt) [reply docs/dev/matrix-event-cache-interaction-contract.md]: Fixed in 4ddf9debb: standardized the contract wording to “labeled,” matching the nearby text and test identifier.

**Comment by basnijholt** [comment]
Independent Claude Code Fable 5 review found one real high-severity cache-contract bug: an opaque encrypted m.annotation reaction was point-only during sync but room-scan reconstruction could add it to a thread snapshot, forcing stale rejection. Verified against the production paths and fixed at the thread-snapshot grouping boundary in 17eb99f5d. Added a focused SQLite/Postgres regression proving refresh stays certified, the reaction is excluded from the snapshot, and no event-to-thread mapping is persisted. Full local gate: Tach passed; every pre-commit hook passed; pytest 10814 passed, 151 skipped.

**Comment by basnijholt** [comment]
@coderabbitai review

**Comment by basnijholt** [comment]
Final independent Fable review on 17eb99f5d found one real blocker: a later clear sync/live/outbound append could overwrite an opaque/full-refetch stale reason, then certify a snapshot still missing ciphertext. Fixed at the durable cache-state boundary in f174da78d: SQLite, PostgreSQL, and the Postgres outage buffer now keep timestamps monotonic while preventing incremental reasons from weakening active full-refetch reasons. Added real SQLite/Postgres policy coverage for all three incremental reason families plus an end-to-end opaque-then-clear sync regression. Full gate: Tach passed; every pre-commit hook passed; pytest 10821 passed, 151 skipped. Remaining Fable observations were critically checked and are non-defects: documented fail-closed/limited-sync choices, unproven latency/assertion risks with intact invariants, inherent undecrypted-metadata limits, and a satisfied SQLite runtime baseline.

**Comment by basnijholt** [comment]
Fresh independent Claude Code Fable 5 max-effort zero-tolerance review completed in an isolated worktree against exact final head `f174da78ddab6ed337c4d487ccdbaaefc50ee7ae`: **APPROVE**, no blocking findings and no source changes. The reviewer independently inspected the complete PR diff and Matrix/cache call paths, exercised real SQLite and PostgreSQL behavior, ran focused suites plus the full repository suite, and passed Tach, Ruff, ty, and import-graph checks. Its seven non-blocking observations were treated as untrusted and re-verified; they are explicit contract choices, safe conservative behavior, satisfied runtime requirements, or unproven overload/robustness risks with current invariants intact.

**Comment by basnijholt** [comment]
Closing as superseded by the four evidence-backed focused replacements: #1589, #1590, #1591, and #1592.

The two final correctness findings were independently verified and fixed:

- Unresolved opaque room-scan impact now invalidates all room thread snapshots in #1591.
- Outbound send/edit/redaction now evicts same-turn thread-history memos in #1592.

The malformed stored-JSON suggestion on #1590 was verified as external database-corruption recovery with no production evidence and was intentionally rejected as speculative scope.

Final combined integration head: `052a93a2156922e6b9b5cd30912aff463de73c8a`.

Verification:

- Full repository pytest passed.
- All pre-commit hooks passed.
- Tach dependency/interface checks passed.
- Every focused PR CI/AI check is green.
- Zero unresolved review threads remain.
- Independent isolated Claude Code Fable 5/max review of the exact integration head: PASS, no blockers.

Production delta:

- #1584: +959/−337 under `src`, net +622 lines.
- Focused integration: +747/−334 under `src`, net +413 lines.

That removes 209 net production lines while retaining the proven fixes and excluding speculative relation/index/corruption machinery.

Nothing was merged.
